### PR TITLE
Fix custom op multi-gpu scaling

### PR DESCRIPTION
**Pull request by piiswrong**
## Description ##
(Brief description on what this PR is about)

### Essentials ###
- [ ] Passed code style checking (`make lint`)
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [ ] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- [ ] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [ ] Feature1, tests, (and when applicable, API doc)
- [ ] Feature2, tests, (and when applicable, API doc)

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

### CODE REVIEW

File: src/operator/custom/custom-inl.h [hunk 1]
```diff
@@ -63,11 +64,80 @@ class Registry {
...
+        bool prev_training = Imperative::Get()->set_is_training(training);
+
+        func();
+
+        Imperative::Get()->set_is_training(prev_training);
+        Imperative::Get()->set_is_recording(prev_recording);
+
+        std::vector<Engine::VarHandle> vars;
+        for (const auto& i : arrs) vars.push_back(i.var());
+        Engine::Get()->PushSync([=](RunContext rctx) {
```
Comment (eric-haibin-lin) [inline src/operator/custom/custom-inl.h]: Will worker thread and main thread push operations to the engine at the same time?

Comment (piiswrong) [reply src/operator/custom/custom-inl.h]: That's fine. Engine is thread safe

**Comment by marcoabreu** [comment]
Just FYI, every GPU-instance has two GPUs present, so feel free to write a test for this case.

File: src/operator/custom/custom-inl.h [hunk 2]
```diff
@@ -63,11 +64,80 @@ class Registry {
...
+    if (std::string("NaiveEngine") != dmlc::GetEnv("MXNET_ENGINE_TYPE", std::string())) {
+      naive_engine_ = false;
+      worker_ = std::thread(
+        [&]() {
+          std::unique_lock<std::mutex> lock(mutex_);
+          while (!q_.empty() || !destructing_) {
+            cv_.wait(lock, [&] {return !q_.empty() || destructing_;});
+            while (!q_.empty()) {
+              auto fn = q_.front();
+              lock.unlock();
```
Comment (anirudh2290) [inline src/operator/custom/custom-inl.h]: What happens when after we unlock, something is pushed to the queue ? Will we pop the correct item from the queue ?

Comment (piiswrong) [reply src/operator/custom/custom-inl.h]: there is on 1 consumer thread, so its ok

**Pull request merged**